### PR TITLE
jsx-first-pro-new-line autofix implementation

### PR DESCRIPTION
**Pull request by snowypowers**
First submission of autofix
- Add fixable to rules file
- Add fixers for each of the 2 cases
- Add output code for each invalid test case

### CODE REVIEW

File: tests/lib/rules/jsx-first-prop-new-line.js [hunk 1]
```diff
@@ -120,6 +120,10 @@ ruleTester.run('jsx-first-prop-new-line', rule, {
   invalid: [
     {
       code: '<Foo prop="one" />',
+      output: [
+        '<Foo',
+        ' prop="one" />'
```
Comment (yangshun) [inline tests/lib/rules/jsx-first-prop-new-line.js]: The indentation looks off (cross ref line 101 and line 111).

Comment (snowypowers) [reply tests/lib/rules/jsx-first-prop-new-line.js]: Hi I have been looking at this problem. To solve this, I would have to change the schema to intake the indent size. Would this be ok?

On a side note, jsx-indent-props solves this properly but would need a second run.

**Comment by snowypowers** [comment]
Indentation has been fixed. A new test case has been added to showcase ability to specify indentation style.

File: tests/lib/rules/jsx-first-prop-new-line.js [hunk 2]
```diff
@@ -120,6 +120,10 @@ ruleTester.run('jsx-first-prop-new-line', rule, {
   invalid: [
     {
       code: '<Foo prop="one" />',
+      output: [
```
Comment (shioju) [inline tests/lib/rules/jsx-first-prop-new-line.js:123]: Pls add test cases for tabs too. Thanks.

Comment (snowypowers) [reply tests/lib/rules/jsx-first-prop-new-line.js:123]: just added.

File: lib/rules/jsx-first-prop-new-line.js [hunk 3]
```diff
@@ -15,14 +15,63 @@ module.exports = {
...
+        }, {
+          type: 'integer'
+        }]
+      }
+    ]
   },
 
   create: function (context) {
     var configuration = context.options[0];
+    var extraColumnStart = 0;
```
Comment (shioju) [inline lib/rules/jsx-first-prop-new-line.js]: This is always zero. Can delete.

File: lib/rules/jsx-first-prop-new-line.js [hunk 4]
```diff
@@ -15,14 +15,63 @@ module.exports = {
...
+        indentSize = context.options[1];
+        indentType = 'space';
+      }
+    }
+
+    function getNodeIndent(node, byLastLine, excludeCommas) {
+      byLastLine = byLastLine || false;
+      excludeCommas = excludeCommas || false;
+
+      var src = sourceCode.getText(node, node.loc.start.column + extraColumnStart);
```
Comment (shioju) [inline lib/rules/jsx-first-prop-new-line.js]: `extraColumnStart` is always zero.

File: lib/rules/jsx-first-prop-new-line.js [hunk 5]
```diff
@@ -35,7 +84,13 @@ module.exports = {
             if (decl.loc.start.line === node.loc.start.line) {
               context.report({
                 node: decl,
-                message: 'Property should be placed on a new line'
+                message: 'Property should be placed on a new line',
+                fix: function(fixer) {
+                  var nodeIndent = getNodeIndent(node, false, false);
```
Comment (shioju) [inline lib/rules/jsx-first-prop-new-line.js]: `getNodeIndent`'s 2nd and 3rd arguments are default to `false`, can remove here.

File: lib/rules/jsx-first-prop-new-line.js [hunk 6]
```diff
@@ -15,14 +15,63 @@ module.exports = {
...
+        indentSize = 1;
+        indentType = 'tab';
+      } else if (typeof context.options[1] === 'number') {
+        indentSize = context.options[1];
+        indentType = 'space';
+      }
+    }
+
+    function getNodeIndent(node, byLastLine, excludeCommas) {
+      byLastLine = byLastLine || false;
```
Comment (shioju) [inline lib/rules/jsx-first-prop-new-line.js]: `getNodeIndent` is only called in one place, and called with `byLastLine` and `excludeCommas` set to false. Can just remove these 2 variables and short circuit their values into this function to make it simpler.

**Review by shioju — CHANGES_REQUESTED** [review]
Can simplify `getNodeIndent` further.

**Comment by snowypowers** [comment]
function amended as requested.

**Review by shioju — APPROVED** [review]

**Comment by shioju** [comment]
We will contact you via email for payment matters. Pls submit a PR upstream as well, for the benefit of everyone else. Thanks. :-)

**Comment by shioju** [comment]
Forgot to mention that you should also update `README.md` to add `(fixable)` next to this rule, and also update `jsx-first-prop-new-line.md` with the identation options. Can refer to `jsx-indent-props.md`.

**Comment by revisionfour** [comment]
Are you guys still working on this? 
[https://github.com/yannickcr/eslint-plugin-react/pull/886](https://github.com/yannickcr/eslint-plugin-react/pull/886)

There are some teams across the globe that could really benefit from a --fix command for this eslint rule. Let me know if you need help. @snowypowers @shioju @yangshun @gyng @lawliet89

**Comment by shioju** [comment]
yes @revisionfour, but right now it can't handle the case where the first line has more than one prop. gonna fix that then push. thanks for asking.

**Pull request merged**